src/realm/object_converter.cpp

```diff
@@ -253,29 +246,189 @@ void InterRealmValueConverter::copy_dictionary(const Obj& src_obj, Obj& dst_obj,
+            handle_dict_copy(i);
+        }
+        else {
+            // copy single element
+            InterRealmValueConverter::ConversionResult converted_src;
+            auto dst_obj = dst_list.get_obj();
+            auto dst_col = dst_list.get_col_key();
+            auto dst_mixed = dst_obj.get_any(dst_col);
+            if (cmp_src_to_dst(any, dst_mixed, &converted_src, update_out)) {
+                dst_list.insert(i, converted_src.converted_value);
             }
         }
     }
 }
 
+void InterRealmValueConverter::handle_dictionary_in_mixed(const Dictionary& src_dict, Dictionary& dst_dict,
+                                                          bool* update_out)
+{
+    // utility functions
+
+    auto handle_set_copy = [&, this](StringData key) {
+        if (!dst_dict.contains(key))
+            dst_dict.insert_collection(key, CollectionType::Set);
+        auto n_src_set = src_dict.get_set(key);
+        auto n_dst_set = dst_dict.get_set(key);
+        copy_set(*n_src_set, *n_dst_set, update_out);
+    };
+    auto handle_list_copy = [&, this](StringData key) {
+        if (!dst_dict.contains(key))
+            dst_dict.insert_collection(key, CollectionType::List);
+        auto n_src_list = src_dict.get_list(key);
+        auto n_dst_list = dst_dict.get_list(key);
+        handle_list_in_mixed(*n_src_list, *n_dst_list, update_out);
+        copy_list(*n_src_list, *n_dst_list, update_out);
+    };
+    auto handle_dict_copy = [&, this](StringData key) {
+        if (!dst_dict.contains(key))
+            dst_dict.insert_collection(key, CollectionType::Dictionary);
+        auto n_src_dict = src_dict.get_dictionary(key);
+        auto n_dst_dict = dst_dict.get_dictionary(key);
+        handle_dictionary_in_mixed(*n_src_dict, *n_dst_dict, update_out);
+        copy_dictionary(*n_src_dict, *n_dst_dict, update_out);
+    };
+
+    // main algorithm
+    auto n = src_dict.size();
+    for (size_t i = 0; i < n; ++i) {
+        auto [key, any] = src_dict.get_pair(i);
+        if (any.is_type(type_List)) {
+            if (dst_dict.contains(key) && check_collection_in_mixed_mismatch(any, dst_dict.get(key), type_List))
+                dst_dict.insert_collection(key.get_string(), CollectionType::List);
+            handle_list_copy(key.get_string());
+        }
+        else if (any.is_type(type_Set)) {
+            if (dst_dict.contains(key) && check_collection_in_mixed_mismatch(any, dst_dict.get(key), type_Set))
+                dst_dict.insert_collection(key.get_string(), CollectionType::Set);
+            handle_set_copy(key.get_string());
+        }
+        else if (any.is_type(type_Dictionary)) {
+            if (dst_dict.contains(key) && check_collection_in_mixed_mismatch(any, dst_dict.get(key), type_Dictionary))
+                dst_dict.insert_collection(key.get_string(), CollectionType::Dictionary);
+            handle_dict_copy(key.get_string());
+        }
+        else {
```

Comment:
is this part necessary? For every callsite of `handle_dictionary_in_mixed()` you also call `copy_dictionary()` which already handles primitive elements.

Reply:
I think we do need this, because of the recursive nature of these collections. For example, you can have a list like:
`[ Mixed{10}, Mixed{"Hello"}, Mixed{ [10, 20, 30] } ];`

`copy_list/copy_set/copy_dictionary` will handle the internal copy of the data held by the generic collections (in this case at position 2). But the other elements won't be treated as collections, thus they need to be copied as well, such as any other scalar/primitive type. And this applies to each level.  So we are not copying twice the same data.

Reply:
If it is not redundant then where is the code is responsible for deleting excess elements from dst? This code here is just inserting all elements from src.

Reply:
I have taken the liberty of adding some failing test to show that the reset isn't behaving as expected. The problem with the current logic is that all elements of the source are _inserted_ into the dest, and then all the original elements are deleted. While this produces the correct end state, it invalidates all user accessors to these objects and creates the wrong notifications. The test I added also uncovered a crash when a nested list has a notification listener and the list is deleted; that will have to be fixed properly so that the notifier is made aware that the collection has been deleted by setting the `collection_root_was_deleted` flag appropriately.

Reply:
`collection_root_was_deleted ` has been fixed already. I have changed a bit the algorithm, but it is still a WIP.